### PR TITLE
throw a cpk warning on the CLI when a schema has CPK, the project has datastore enabled and the corresponding feature flag is disabled

### PR DESCRIPTION
**Pull request by manaswi223**
#### Description of changes

This PR helps the CX by throwing a CPK warning ⚠️ on the CLI when:
1. schema has a CPK - @primarykey and
2. conflict detection is enabled (datastore is used in the project) and
3. the feature flag for CPK - `respectPrimaryKeyAttributesOnConnectionField` is false in "amplify/cli.json" file

When CPK modelgen is being disabled in a schema with CPK's, this PR would aid the CX to enable the feature flag and solve issues with the CPK models.

sample schema used:

```
type Todo @model {
  todoId: ID! @primaryKey
  content: String
}

type TodoCopy @model {
  todoCopyId: ID! @primaryKey
  contentCopy: String
}
```

screenshot of the warning when the schema is compiled by running the command -
`amplify-dev api gql-compile`

Screen Shot 2022-12-28 at 12 23 57 PM

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [x] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-graphql-transformer-core/src/transformation/transform.ts [hunk 1]
```diff
@@ -822,4 +825,17 @@ export class GraphQLTransform {
...
     }
   }
+
+// Checks if DataStore is enabled and CPK Feature Flag is true. Throws a warning otherwise.
+public validateCPKFeatureFlag = (context: TransformerContext ) => {
+  const isDataStoreEnabled = context.isProjectUsingDataStore();
+  if (isDataStoreEnabled && !isCPKFeatureEnabled(context) && context.isProjectUsingCPK()) {
+    printer.info(
+      `
+⚠️  WARNING: Your schema has a custom primary key but the Feature Flag "${cpkFeatureFlagName}" is disabled. Check the value in your "amplify/cli.json" file, change it to "true" and re-run
```
Comment (david-mcafee) [inline packages/amplify-graphql-transformer-core/src/transformation/transform.ts:835]: I would clarify what we mean by `re-run` here. For example:

```suggestion
⚠️  WARNING: Your schema has a custom primary key but the Feature Flag "${cpkFeatureFlagName}" is disabled. Check the value in your "amplify/cli.json" file, change it to `true`, and execute `amplify codegen models` with the CLI
```

@renebrandel any thoughts on the wording, here? Thanks!

**Review by david-mcafee — CHANGES_REQUESTED** [review]
Left a minor suggestion

File: packages/amplify-graphql-transformer-core/src/transformer-context/index.ts [hunk 2]
```diff
@@ -98,4 +101,12 @@ export class TransformerContext implements TransformerContextProvider {
   public isProjectUsingDataStore(): boolean {
     return !!this.resolverConfig?.project || !!this.resolverConfig?.models;
   }
+
+  public isProjectUsingCPK = (): boolean => {
```
Comment (david-mcafee) [inline packages/amplify-graphql-transformer-core/src/transformer-context/index.ts:105]: Can we add unit tests for this check? Thanks!!

**Review by david-mcafee — CHANGES_REQUESTED** [review]

File: packages/amplify-graphql-transformer-core/src/transformation/transform.ts (hunk 1 above)
Comment (manaswi223) [reply packages/amplify-graphql-transformer-core/src/transformation/transform.ts:835]: Another CX:
For some customers using the old Amplify projects -->
manually switching the FF to true can also be a breaking change and might lose some data. Should we also warn them regarding this? 

thank you @AaronZyLee for catching this perception

Comment (david-mcafee) [reply packages/amplify-graphql-transformer-core/src/transformation/transform.ts:835]: @AaronZyLee - can you elaborate more on this?

**Comment by david-mcafee** [comment]
@manaswi223 looks like this PR can be closed now, since [the other PR](https://github.com/aws-amplify/amplify-category-api/pull/1126) was merged?

**Comment by manaswi223** [comment]
no @david-mcafee David, that PR is closed and this is the actual PR & awaiting review from @renebrandel.

File: packages/amplify-graphql-transformer-core/src/transformation/transform.ts (hunk 1 above)
Comment (david-mcafee) [reply packages/amplify-graphql-transformer-core/src/transformation/transform.ts:835]: @manaswi223 - thoughts on the suggestion I made here (i.e. asking the customer to re-run `amplify codegen models`?

**Comment by david-mcafee** [comment]
> no @david-mcafee David, that PR is closed and this is the actual PR & awaiting review from @renebrandel.

Thanks @manaswi223! I left a suggestion on updating the messaging above :)

**Comment by manaswi223** [comment]
thanks David. Committed that suggestion :)

File: packages/amplify-graphql-transformer-core/src/transformer-context/index.ts [hunk 3]
```diff
@@ -98,4 +101,12 @@ export class TransformerContext implements TransformerContextProvider {
...
+    const obj = this.inputDocument.definitions.find(def => def.kind === 'ObjectTypeDefinition') as ObjectTypeDefinitionNode;
+    const primaryKeyField = obj.fields?.find(field => field.directives?.find(directive => directive.name.value === 'primaryKey'));
+    if (!primaryKeyField || primaryKeyField.name.value === DEFAULT_ID_FIELD_NAME) {
```
Comment (AaronZyLee) [inline packages/amplify-graphql-transformer-core/src/transformer-context/index.ts:108]: I think the following schema should also be considered as custom primary key (id as primary key and with sort key fields defined)
```GraphQL
type Todo @model {
  id: ID! @primaryKey(sortKeyFields:["name"])
  name: String!
}
```

**Review by AaronZyLee — CHANGES_REQUESTED** [review]

File: packages/amplify-graphql-transformer-core/src/transformer-context/index.ts (hunk 2 above)
Comment (david-mcafee) [reply packages/amplify-graphql-transformer-core/src/transformer-context/index.ts:105]: Looks like we're only mocking the return value for this in the unit tests - thoughts on adding testing for this function?

File: packages/amplify-graphql-transformer-core/src/transformer-context/index.ts (hunk 3 above)
Comment (david-mcafee) [reply packages/amplify-graphql-transformer-core/src/transformer-context/index.ts:108]: +1

**Review by david-mcafee — CHANGES_REQUESTED** [review]

**Pull request closed without merging**